### PR TITLE
Added OptOut API to SMS service 2024-12-10-preview

### PR DESCRIPTION
**Pull request by phermanov-msft**
# Data Plane API Specification Update Pull Request

> [!TIP]
> Overwhelmed by all this guidance? See the `Getting help` section at the bottom of this PR description.

The purpose of the opt-out management API is to enable customers to manage opt-out requests for SMS messages. The Opt-Out Management API is a self-service API that enables businesses to manage opt-out requests for their SMS campaigns.

## PR review workflow diagram

Please understand this diagram before proceeding. It explains how to get your PR approved & merged.

spec_pr_review_workflow_diagram

## API Info: The Basics

Most of the information about your service should be captured in the issue that serves as your [*API Spec engagement record*](https://aka.ms/azsdk/onboarding/restapischedule).

* Link to API Spec engagement record issue: https://github.com/Azure/azure-rest-api-specs/issues/30948

Is this review for (select one):

- [ ] a private preview
- [x] a public preview
- [ ] GA release

### Change Scope

<sup>This section will help us focus on the specific parts of your API that are new or have been modified. <br/>Please share a link to the design document for the new APIs, a link to the previous API Spec document (if applicable), and the root paths that have been updated. </sup>
* Design Document: https://github.com/Azure/communication-preview/blob/master/SMS/Opt-out-management-concept.md
* Previous API Spec Doc: https://github.com/Azure/azure-rest-api-specs/blob/main/specification/communication/data-plane/Sms/preview/2024-02-05-preview/communicationservicessms.json
* Updated paths:
  - /sms/optouts/:add
  - /sms/optouts/:remove
  - /sms/optouts/:check
* Glossary: https://skype.visualstudio.com/SPOOL/_wiki/wikis/SPOOL.wiki/56990/Glossary

### Viewing API changes

For convenient view of the API changes made by this PR, refer to the URLs provided in the table 
in the `Generated ApiView` comment added to this PR. You can use ApiView to show API versions diff. 

### Suppressing failures

If one or multiple validation error/warning suppression(s) is detected in your PR, please follow the 
[Swagger-Suppression-Process](https://aka.ms/pr-suppressions) 
to get approval.

## ❔Got questions? Need additional info?? We are here to help!

<details>
  <summary> Contact us!</summary>

The [Azure API Review Board](https://aka.ms/azsdk/onboarding/restapischedule) is dedicated to helping you create amazing APIs. You can read about our mission and learn more about our process on our [wiki](https://aka.ms/azsdk/onboarding/restapischedule).
* 💬 [Teams Channel](https://teams.microsoft.com/l/channel/19%3a3ebb18fded0e47938f998e196a52952f%40thread.tacv2/General?groupId=1a10b50c-e870-4fe0-8483-bf5542a8d2d8&tenantId=72f988bf-86f1-41af-91ab-2d7cd011db47)
* 💌 [email](mailto://azureapirbcore@microsoft.com)

</details>

<details>
  <summary>Click here for links to tools, specs, guidelines & other good stuff</summary>

### Tooling

 * [Open API validation tools](https://aka.ms/swaggertools) were run on this PR. Go here to see [how to fix errors](https://aka.ms/ci-fix)
 * [Spectral Linting](https://github.com/Azure/azure-api-style-guide/blob/main/README.md)

### Guidelines & Specifications

 * [Azure REST API Guidelines](https://aka.ms/azapi/guidelines)
 * [OpenAPI Style Guidelines](https://aka.ms/azapi/style)
 * [Azure Breaking Change Policy](https://aka.ms/AzBreakingChangesPolicy)

### Helpful Links

 * [Schedule a data plane REST API spec review](https://aka.ms/azsdk/onboarding/restapischedule)

</details>

## Getting help

- First, please carefully read through this PR description, from top to bottom.
- If you don't have permissions to remove or add labels to the PR, request `write access` per [aka.ms/azsdk/access#request-access-to-rest-api-or-sdk-repositories](https://aka.ms/azsdk/access#request-access-to-rest-api-or-sdk-repositories) 
- To understand what you must do next to merge this PR, see the `Next Steps to Merge` comment. It will appear within few minutes of submitting this PR and will continue to be up-to-date with current PR state.
- For guidance on fixing this PR CI check failures, see the hyperlinks provided in given failure 
  and https://aka.ms/ci-fix.
- If the PR CI checks appear to be stuck in `queued` state, please add a comment with contents `/azp run`.
  This should result in a new comment denoting a `PR validation pipeline` has started and the checks should be updated after few minutes.
- If the help provided by the previous points is not enough, post to https://aka.ms/azsdk/support/specreview-channel and link to this PR.

### CODE REVIEW

File: specification/communication/data-plane/Sms/preview/2024-12-10-preview/communicationservicessms.json [hunk 1]
```diff
@@ -62,6 +62,138 @@
           }
         }
       }
+    },
+    "/sms/optouts/:add": {
+      "post": {
+        "tags": [
+          "OptOutsAdd"
+        ],
+        "summary": "Allows Contoso's users to opt out (unsubscribe) from a phone number and block receiving messages from that number.",
```
Comment (DominikMe) [inline specification/communication/data-plane/Sms/preview/2024-12-10-preview/communicationservicessms.json:71]: Let's avoid the term "Contoso" in public docs.

```suggestion
        "summary": "Add phone numbers to the optouts list which shall stop receiving messages from a sender number.",
```

File: specification/communication/data-plane/Sms/preview/2024-12-10-preview/communicationservicessms.json [hunk 2]
```diff
@@ -62,6 +62,138 @@
...
+          }
+        }
+      }
+    },
+    "/sms/optouts/:remove": {
+      "post": {
+        "tags": [
+          "OptOutsRemove"
+        ],
+        "summary": "Allows Contoso's users to opt in (subscribe) to a phone number and enable receiving messages from that number.",
```
Comment (DominikMe) [inline specification/communication/data-plane/Sms/preview/2024-12-10-preview/communicationservicessms.json:116]:
```suggestion
        "summary": "Remove phone numbers from the optouts list.",
```

File: specification/communication/data-plane/Sms/preview/2024-12-10-preview/communicationservicessms.json [hunk 3]
```diff
@@ -62,6 +62,138 @@
...
+          }
+        }
+      }
+    },
+    "/sms/optouts/:check": {
+      "post": {
+        "tags": [
+          "OptOutsCheck"
+        ],
+        "summary": "Allows Contoso's users to check opt status for a phone number if it can receive messages from that number.",
```
Comment (DominikMe) [inline specification/communication/data-plane/Sms/preview/2024-12-10-preview/communicationservicessms.json:161]:
```suggestion
        "summary": "Check the opt out status for a recipient phone number with a sender phone number.",
```

File: specification/communication/data-plane/Sms/preview/2024-12-10-preview/communicationservicessms.json [hunk 4]
```diff
@@ -197,6 +329,84 @@
           "type": "string"
         }
       }
+    },
+    "OptOutSendRecipient": {
+      "description": "Represents properties for a single recipient.",
```
Comment (DominikMe) [inline specification/communication/data-plane/Sms/preview/2024-12-10-preview/communicationservicessms.json:337]:
```suggestion
      "description": "A single opt out recipient number.",
```

File: specification/communication/data-plane/Sms/preview/2024-12-10-preview/communicationservicessms.json [hunk 5]
```diff
@@ -197,6 +329,84 @@
...
+      "properties": {
+        "to": {
+          "description": "The recipient's phone number (in E.164 format).",
+          "minLength": 1,
+          "type": "string"
+        }
+      }
+    },
+    "OptOutSendRequest": {
+      "description": "Represents properties for a Opt request.",
```
Comment (DominikMe) [inline specification/communication/data-plane/Sms/preview/2024-12-10-preview/communicationservicessms.json:351]:
```suggestion
      "description": "An opt out request.",
```

File: specification/communication/data-plane/Sms/preview/2024-12-10-preview/communicationservicessms.json [hunk 6]
```diff
@@ -197,6 +329,84 @@
...
+          "maxItems": 100,
+          "minItems": 1,
+          "type": "array",
+          "items": {
+            "$ref": "#/definitions/OptOutSendRecipient"
+          }
+        }
+      }
+    },
+    "OptOutSendResponse": {
```
Comment (DominikMe) [inline specification/communication/data-plane/Sms/preview/2024-12-10-preview/communicationservicessms.json:373]: Consider dropping the Send in most of the names if you feel like `OptOutResponse`, `OptOutRecipient` etc. is descriptive enough to be understood. I think you mean that ((OptOutSend)Response) means opting out of sending, but I personally trip up when I read SendResponse and SendRecipient next to each other.

File: specification/communication/data-plane/Sms/preview/2024-12-10-preview/communicationservicessms.json [hunk 7]
```diff
@@ -197,6 +329,84 @@
...
+      "description": "Response for a successful or multi status Opt send request.",
```
Comment (DominikMe) [inline specification/communication/data-plane/Sms/preview/2024-12-10-preview/communicationservicessms.json:374]: "multi status" implies a 207 response, let's avoid this term and rephrase to something like
```suggestion
      "description": "Response for an opt out request. Validate the returned items in the response to see which recipients were successfully added or removed from the opt outs list.",
```

**Review by DominikMe — COMMENTED** [review]
Some comments on description messages, overall looking good

File: specification/communication/data-plane/Sms/preview/2024-12-10-preview/communicationservicessms.json (hunk 6 above)
Comment (phermanov-msft) [reply specification/communication/data-plane/Sms/preview/2024-12-10-preview/communicationservicessms.json]: Renamed the difinitions

**Review by DominikMe — APPROVED** [review]
Lgtm for preview

**Pull request merged**